### PR TITLE
Implement sidecar tasks and task resource specification

### PR DESCRIPTION
**Pull request by katrogan**
Moved the resources to a separate location to avoid a circular dependency with task and python_function_task, let me know if it belongs elsewhere.

Also the foo().with_overrides(cpu="1") syntax is not implemented by these changes

### CODE REVIEW

File: flytekit/taskplugins/sidecar/task.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+from google.protobuf.json_format import MessageToDict
+
+from flytekit.annotated.context_manager import FlyteContext, RegistrationSettings
+from flytekit.annotated.promise import Promise
+from flytekit.annotated.python_function_task import PythonFunctionTask
+from flytekit.annotated.task import TaskPlugins
+from flytekit.common.exceptions import user as _user_exceptions
+from flytekit.models import task as _task_model
+from flytekit.models import task as _task_models
+from flytekit.plugins import k8s as _lazy_k8s
```
Comment (kumare3) [inline flytekit/taskplugins/sidecar/task.py:13]: I do not think you need to do this. But, I guess lets keep it. I need to move all the _lazy_k8s logic to the plugins

File: flytekit/annotated/resources.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    storage: str = None
+
+
+@dataclass
+class ResourceSpec(object):
+    requests: Resource = None
+    limits: Resource = None
+
+
+def _get_resources(**kwargs):
```
[the remark this answers is not in the document]
Comment (katrogan) [reply flytekit/annotated/resources.py:18]: i will gladly not _

File: flytekit/annotated/python_function_task.py [hunk 3]
```diff
@@ -88,6 +89,8 @@ def __init__(
         self._task_function = task_function
         self._task_config = task_config
         self._container_image = container_image
+        # TODO(katrogan): Implement resource overrides
+        self._resources = _get_resources(**kwargs)
```
[the remark this answers is not in the document]
Comment (katrogan) [reply flytekit/annotated/python_function_task.py]: done

File: flytekit/annotated/python_function_task.py [hunk 4]
```diff
@@ -125,4 +132,12 @@ def get_container(self, settings: RegistrationSettings) -> _task_model.Container
...
             environment=env,
+            storage_request=self._resources.requests.storage,
+            cpu_request=self._resources.requests.cpu,
+            gpu_request=self._resources.requests.gpu,
+            memory_request=self._resources.requests.mem,
+            storage_limit=self._resources.limits.storage,
+            cpu_limit=self._resources.limits.cpu,
+            gpu_limit=self._resources.limits.gpu,
+            memory_limit=self._resources.limits.mem,
         )
```
Comment (wild-endeavor) [inline flytekit/annotated/python_function_task.py:143]: should add `**kwargs` as the last item.

Comment (katrogan) [reply flytekit/annotated/python_function_task.py:143]: talked offline, added to get_resources

**Pull request merged**